src/Lean/Elab/Match.lean

```diff
@@ -22,6 +22,17 @@ namespace Lean.Elab.Term
 open Meta
 open Lean.Parser.Term
 
+/--
+Determines the local declaration kind depending on the variable name.
+
+The `__x` in `let __x := 42; body` gets kind `.implDetail`.
+-/
+def kindOfBinderName (binderName : Name) : LocalDeclKind :=
+  if binderName.isImplementationDetail then
+    .implDetail
+  else
+    .default
```

Comment:
Can we keep this in its original position and have Lean.Elab.Match import Lean.Elab.Binders? This isn't the natural place for this function.

Reply:
How about `Lean.Elab.BindersUtil`? `Match` already imports that and this is for sure a binder util.

Reply:
In the course of creating a test, I observed that `intro` also doesn't respect the convention, so I added `ofBinderName` there as well. But `Meta.Tactic.Intro` doesn't import `BindersUtil`. In the end I just put the utility in `LocalContext`. It's a more fundamental file so feels scary to change, but hopefully this makes sense. Open to suggestions of other utility modules to put it in (I couldn't find any ones that fit).

Reply:
I'm unsure about having `intro` respect this, since if somehow you have `let __x := 1; p __x` and do `intro`, you'll get the goal `p __x` without any `__x` in the local context. Seems unfriendly. I could see, potentially, `intro __x` with an explicitly provided name doing this, but how about, given that there are no applications of this yet, we avoid modifying `intro` for now?

I would suggest that we keep kindOfBinderName in BindersUtil since it seems like an elaboration concern to me, not a meta-level concern. Might be reasonable to keep it as `LocalDeclKind.ofBinderName`. Could you change the docstring to something clearer while we're here? Like "Uses the binder name to choose the local declaration kind. Names that begin with `__` are implementation details."

Reply:
Ok, done. In my mind `intro` is just funny syntax for `let` or `fun` so treating them differently is strange; but I agree there's no need to change it without a usecase in mind.